File: typescript/src/autonomous-action-admission-controller.ts
import { ArgumentError, isObject } from "./errors.js";
import { AUTONOMOUS_DOMAIN_NAMES, type AutonomousDomainName } from "./autonomous.js";
import {
  AutonomousActionAdmission,
  AutonomousActionPlan,
  type AutonomousActionAdmissionJSON,
  type AutonomousActionPlanJSON,
  admitAutonomousActionPlan,
} from "./autonomous-action-plan.js";
import {
  InMemoryAutonomousActionAdmissionLedger,
  validateAutonomousActionAdmissionRecord,
  type AutonomousActionAdmissionRecord,
  type AutonomousActionAdmissionReviewOptions,
} from "./autonomous-action-admission-persistence.js";
import { digestJsonSync } from "./tooling.js";
import type { JsonObject } from "./types.js";
import type { AutonomousActionPlanApproval } from "./autonomous-action-plan.js";

/** Bounded operator review projection; it never includes the task or a runtime payload. */
export const AUTONOMOUS_ACTION_REVIEW_QUEUE_SCHEMA = "bioprism-typescript-autonomous-action-review-queue/0.1" as const;
export const AUTONOMOUS_ACTION_REVIEW_ROW_SCHEMA = "bioprism-typescript-autonomous-action-review-row/0.1" as const;
export const AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA = "bioprism-typescript-autonomous-action-dispatch-handoff/0.1" as const;
export const AUTONOMOUS_ACTION_REVIEW_RETENTION = "metadata_only;operator_review_projection_and_digests;task_prompt_provider_connector_credential_and_effect_values_not_retained" as const;
export const AUTONOMOUS_ACTION_REVIEW_AUTHORITY = "caller_operator_projection_only;authorization_is_external_and_not_verified_by_sdk" as const;
export const AUTONOMOUS_ACTION_REVIEW_EXECUTION = "review_control_only;does_not_authorize_provider_source_tool_effect_or_credentials" as const;
export const AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL = "never_returned" as const;
export const AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES = ["credential_scope", "provider_or_source_approval", "tool_and_effect_authority", "evaluator_settlement"] as const;

export interface AutonomousActionReviewRow extends JsonObject {
  schema: typeof AUTONOMOUS_ACTION_REVIEW_ROW_SCHEMA;
  action_id: string;
  revision: number;
  status: "pending_review" | "admitted" | "blocked";
  plan_digest: string;
  admission_digest: string;
  route_digest: string;
  selected_domains: AutonomousDomainName[];
  cross_domain: boolean;
  execution_path: string;
  next_action: string;
  next_actions: string[];
  required_approvals: string[];
  approved_approvals: string[];
  missing_approvals: string[];
  review_reasons: string[];
  blocking_reasons: string[];
  reviewer_digest: string | null;
  reason_digest: string | null;
  record_digest: string;
  authority: typeof AUTONOMOUS_ACTION_REVIEW_AUTHORITY;
  retention: typeof AUTONOMOUS_ACTION_REVIEW_RETENTION;
  execution: typeof AUTONOMOUS_ACTION_REVIEW_EXECUTION;
  secret_material: typeof AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL;
}

export interface AutonomousActionReviewQueue extends JsonObject {
  schema: typeof AUTONOMOUS_ACTION_REVIEW_QUEUE_SCHEMA;
  rows: AutonomousActionReviewRow[];
  counts: {
    total: number;
    pending_review: number;
    admitted: number;
    blocked: number;
  };
  domain_counts: Record<AutonomousDomainName, number>;
  queue_digest: string;
  authority: typeof AUTONOMOUS_ACTION_REVIEW_AUTHORITY;
  retention: typeof AUTONOMOUS_ACTION_REVIEW_RETENTION;
  execution: typeof AUTONOMOUS_ACTION_REVIEW_EXECUTION;
  secret_material: typeof AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL;
}

export interface AutonomousActionDispatchHandoff extends JsonObject {
  schema: typeof AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA;
  action_id: string;
  record_digest: string;
  plan_digest: string;
  admission_digest: string;
  plan: AutonomousActionPlanJSON;
  admission: AutonomousActionAdmissionJSON;
  selected_domains: AutonomousDomainName[];
  requested_domains: AutonomousDomainName[];
  cross_domain: boolean;
  execution_path: string;
  status: "ready_for_downstream_gates";
  downstream_gates: string[];
  authority: typeof AUTONOMOUS_ACTION_REVIEW_AUTHORITY;
  retention: typeof AUTONOMOUS_ACTION_REVIEW_RETENTION;
  execution: typeof AUTONOMOUS_ACTION_REVIEW_EXECUTION;
  secret_material: typeof AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL;
  handoff_digest: string;
}

export interface AutonomousActionOperatorReviewOptions extends Omit<AutonomousActionAdmissionReviewOptions, "reviewerDigest"> {
  authorizationDigest: string;
}

export interface AutonomousActionOperatorSubmitOptions {
  approvals?: Partial<Record<AutonomousActionPlanApproval, boolean>>;
  reviewed?: boolean;
  authorizationDigest?: string | null;
  reason?: string | null;
}

function fail(message: string): never {
  throw new ArgumentError(`autonomous action review controller ${message}`);
}

function digest(name: string, value: unknown, allowNull = false): string | null {
  if (allowNull && value === null) return null;
  if (typeof value !== "string" || !/^[0-9a-f]{64}$/.test(value)) fail(`${name} must be a lowercase SHA-256 digest`);
  return value;
}

function domainList(name: string, value: readonly string[]): AutonomousDomainName[] {
  if (!Array.isArray(value) || value.length < 1 || value.length > AUTONOMOUS_DOMAIN_NAMES.length) fail(`${name} must contain one to twelve domains`);
  const result = value.map((domain) => {
    if (!AUTONOMOUS_DOMAIN_NAMES.includes(domain as AutonomousDomainName)) fail(`${name} contains an unsupported domain`);
    return domain as AutonomousDomainName;
  });
  if (new Set(result).size !== result.length) fail(`${name} must contain unique domains`);
  return result;
}

const DISPATCH_HANDOFF_KEYS = [
  "schema", "action_id", "record_digest", "plan_digest", "admission_digest", "plan", "admission",
  "selected_domains", "requested_domains", "cross_domain", "execution_path", "status", "downstream_gates",
  "authority", "retention", "execution", "secret_material", "handoff_digest",
] as const;

function exactHandoffKeys(value: Record<string, unknown>): void {
  const allowed = new Set<string>(DISPATCH_HANDOFF_KEYS);
  if (Object.keys(value).some((key) => !allowed.has(key)) || DISPATCH_HANDOFF_KEYS.some((key) => !(key in value))) fail("handoff has unsupported or missing fields");
}

function actionId(value: unknown): string {
  if (typeof value !== "string" || value.length < 1 || value.length > 256 || !/^[A-Za-z0-9_.:+/-]+$/.test(value)) fail("handoff action_id is invalid");
  return value;
}

/**
 * Verify a handoff after it has crossed a persistence or process boundary.
 *
 * This proves only metadata continuity. It deliberately does not prove reviewer authorization,
 * credential availability, provider readiness, source truth, evaluator quality, or effect safety;
 * those remain independent downstream gates.
 */
export function validateAutonomousActionDispatchHandoff(value: unknown): AutonomousActionDispatchHandoff {
  if (!isObject(value)) fail("handoff must be an object");
  exactHandoffKeys(value);
  if (value.schema !== AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA || value.authority !== AUTONOMOUS_ACTION_REVIEW_AUTHORITY || value.retention !== AUTONOMOUS_ACTION_REVIEW_RETENTION || value.execution !== AUTONOMOUS_ACTION_REVIEW_EXECUTION || value.secret_material !== AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL) fail("handoff markers are invalid");
  const plan = AutonomousActionPlan.fromJSON(value.plan as AutonomousActionPlanJSON);
  const admission = AutonomousActionAdmission.fromJSON(value.admission as AutonomousActionAdmissionJSON);
  const recordDigest = digest("handoff record_digest", value.record_digest) as string;
  const planDigest = digest("handoff plan_digest", value.plan_digest) as string;
  const admissionDigest = digest("handoff admission_digest", value.admission_digest) as string;
  const suppliedHandoffDigest = digest("handoff handoff_digest", value.handoff_digest) as string;
  if (planDigest !== plan.plan_digest || admissionDigest !== admission.admission_digest || plan.plan_digest !== admission.plan_digest) fail("handoff plan or admission digest is inconsistent");
  if (admission.status !== "admitted") fail("handoff admission is not admitted");
  const selected = domainList("handoff selected_domains", value.selected_domains as readonly string[]);
  const requested = domainList("handoff requested_domains", value.requested_domains as readonly string[]);
  if (JSON.stringify(selected) !== JSON.stringify(plan.selected_domains) || JSON.stringify(selected) !== JSON.stringify(admission.selected_domains)) fail("handoff selected domains are inconsistent");
  if (requested.some((domain) => !selected.includes(domain))) fail("handoff requested domains exceed the selected domains");
  if (typeof value.cross_domain !== "boolean" || value.cross_domain !== plan.cross_domain) fail("handoff cross-domain posture is inconsistent");
  if (typeof value.execution_path !== "string" || value.execution_path !== admission.execution_path) fail("handoff execution path is inconsistent");
  if (value.status !== "ready_for_downstream_gates") fail("handoff status is invalid");
  if (!Array.isArray(value.downstream_gates) || JSON.stringify(value.downstream_gates) !== JSON.stringify(AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES)) fail("handoff downstream gates are invalid");
  const body = {
    schema: AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA,
    action_id: actionId(value.action_id),
    record_digest: recordDigest,
    plan_digest: planDigest,
    admission_digest: admissionDigest,
    plan: plan.toJSON(),
    admission: admission.toJSON(),
    selected_domains: selected,
    requested_domains: requested,
    cross_domain: value.cross_domain,
    execution_path: value.execution_path,
    status: "ready_for_downstream_gates" as const,
    downstream_gates: [...AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES],
    authority: AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
    retention: AUTONOMOUS_ACTION_REVIEW_RETENTION,
    execution: AUTONOMOUS_ACTION_REVIEW_EXECUTION,
    secret_material: AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
  };
  const expected = digestJsonSync(body);
  if (suppliedHandoffDigest !== expected) fail("handoff digest does not match metadata");
  return { ...body, handoff_digest: expected };
}

function normalizedRecord(record: AutonomousActionAdmissionRecord): { record: AutonomousActionAdmissionRecord; plan: AutonomousActionPlan; admission: AutonomousActionAdmission } {
  const normalized = validateAutonomousActionAdmissionRecord(record);
  const plan = AutonomousActionPlan.fromJSON(normalized.plan as AutonomousActionPlanJSON);
  const admission = AutonomousActionAdmission.fromJSON(normalized.admission as AutonomousActionAdmissionJSON);
  return { record: normalized, plan, admission };
}

function rowFor(record: AutonomousActionAdmissionRecord): AutonomousActionReviewRow {
  const { record: normalized, plan, admission } = normalizedRecord(record);
  return {
    schema: AUTONOMOUS_ACTION_REVIEW_ROW_SCHEMA,
    action_id: normalized.action_id,
    revision: normalized.revision,
    status: normalized.status,
    plan_digest: plan.plan_digest,
    admission_digest: admission.admission_digest,
    route_digest: plan.route_digest,
    selected_domains: [...plan.selected_domains],
    cross_domain: plan.cross_domain,
    execution_path: admission.execution_path,
    next_action: admission.next_action,
    next_actions: [...admission.next_actions],
    required_approvals: [...admission.required_approvals],
    approved_approvals: [...admission.approved_approvals],
    missing_approvals: [...admission.missing_approvals],
    review_reasons: [...admission.review_reasons],
    blocking_reasons: [...admission.blocking_reasons],
    reviewer_digest: normalized.reviewer_digest,
    reason_digest: normalized.reason_digest,
    record_digest: normalized.record_digest,
    authority: AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
    retention: AUTONOMOUS_ACTION_REVIEW_RETENTION,
    execution: AUTONOMOUS_ACTION_REVIEW_EXECUTION,
    secret_material: AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
  };
}

function queueFor(rows: AutonomousActionReviewRow[]): AutonomousActionReviewQueue {
  const sorted = [...rows].sort((left, right) => left.action_id.localeCompare(right.action_id));
  const domainCounts = Object.fromEntries(AUTONOMOUS_DOMAIN_NAMES.map((domain) => [domain, 0])) as Record<AutonomousDomainName, number>;
  for (const row of sorted) for (const domain of row.selected_domains) domainCounts[domain] += 1;
  const counts = {
    total: sorted.length,
    pending_review: sorted.filter((row) => row.status === "pending_review").length,
    admitted: sorted.filter((row) => row.status === "admitted").length,
    blocked: sorted.filter((row) => row.status === "blocked").length,
  };
  const body = {
    schema: AUTONOMOUS_ACTION_REVIEW_QUEUE_SCHEMA,
    rows: sorted,
    counts,
    domain_counts: domainCounts,
    authority: AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
    retention: AUTONOMOUS_ACTION_REVIEW_RETENTION,
    execution: AUTONOMOUS_ACTION_REVIEW_EXECUTION,
    secret_material: AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
  } as Omit<AutonomousActionReviewQueue, "queue_digest">;
  return { ...body, queue_digest: digestJsonSync(body) } as AutonomousActionReviewQueue;
}

export class AutonomousActionAdmissionController {
  readonly ledger: InMemoryAutonomousActionAdmissionLedger;

  constructor(ledger: InMemoryAutonomousActionAdmissionLedger) {
    if (!(ledger instanceof InMemoryAutonomousActionAdmissionLedger)) fail("requires a typed action admission ledger");
    this.ledger = ledger;
  }

  queue(): AutonomousActionReviewQueue {
    return queueFor(this.ledger.list().map(rowFor));
  }

  get(actionId: string): AutonomousActionReviewRow | null {
    const record = this.ledger.get(actionId);
    return record === null ? null : rowFor(record);
  }

  /**
   * Submit one provider-free plan to the durable review queue.
   *
   * A submitter may attach a caller authorization digest only when the plan is already
   * admitted; ordinary submissions remain pending until the separate review operation.
   */
  submit(actionId: string, source: AutonomousActionPlan | AutonomousActionPlanJSON, options: AutonomousActionOperatorSubmitOptions = {}): AutonomousActionReviewRow {
    if (!options || typeof options !== "object") fail("submit options are malformed");
    const plan = source instanceof AutonomousActionPlan ? source : AutonomousActionPlan.fromJSON(source);
    const admission = admitAutonomousActionPlan(plan, { approvals: options.approvals, reviewed: options.reviewed ?? false });
    const authorizationDigest = options.authorizationDigest === undefined || options.authorizationDigest === null
      ? null
      : digest("authorizationDigest", options.authorizationDigest);
    if (admission.status === "admitted" && authorizationDigest === null) fail("an admitted submission requires an authorizationDigest");
    const record = this.ledger.submit(plan, admission, {
      actionId,
      reviewerDigest: authorizationDigest,
      reason: options.reason,
    });
    return rowFor(record);
  }

  review(actionId: string, options: AutonomousActionOperatorReviewOptions): AutonomousActionReviewRow {
    if (!options || typeof options !== "object") fail("review options are malformed");
    const authorizationDigest = digest("authorizationDigest", options.authorizationDigest);
    const { authorizationDigest: _authorizationDigest, ...reviewOptions } = options;
    const record = this.ledger.review(actionId, { ...reviewOptions, reviewerDigest: authorizationDigest as string });
    return rowFor(record);
  }

  dispatchHandoff(actionId: string, requestedDomains?: readonly AutonomousDomainName[]): AutonomousActionDispatchHandoff {
    const record = this.ledger.get(actionId);
    if (record === null) fail("cannot create a handoff for an unknown action");
    const { record: normalized, plan, admission } = normalizedRecord(record);
    if (normalized.status !== "admitted" || admission.status !== "admitted") fail("action admission is not ready for downstream gates");
    const selected = [...plan.selected_domains];
    const requested = domainList("requestedDomains", requestedDomains === undefined ? selected : requestedDomains);
    const missing = requested.filter((domain) => !selected.includes(domain));
    if (missing.length) fail(`requested domains are outside the admitted plan: ${missing.join(", ")}`);
    const body = {
      schema: AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA,
      action_id: normalized.action_id,
      record_digest: normalized.record_digest,
      plan_digest: plan.plan_digest,
      admission_digest: admission.admission_digest,
      plan: plan.toJSON(),
      admission: admission.toJSON(),
      selected_domains: selected,
      requested_domains: requested,
      cross_domain: plan.cross_domain,
      execution_path: admission.execution_path,
      status: "ready_for_downstream_gates" as const,
      downstream_gates: [...AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES],
      authority: AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
      retention: AUTONOMOUS_ACTION_REVIEW_RETENTION,
      execution: AUTONOMOUS_ACTION_REVIEW_EXECUTION,
      secret_material: AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
    };
    return validateAutonomousActionDispatchHandoff({ ...body, handoff_digest: digestJsonSync(body) });
  }
}

File: python/prism_sdk/autonomous_action_admission_controller.py
"""Operator-facing review and dispatch-handoff projections for action admissions.

The controller is intentionally narrower than an execution engine. It gives an application a
bounded queue view, exact-record review operation, and downstream handoff. Authorization is
represented by a caller-supplied digest and must be verified by the deployment's identity system;
the controller never grants provider, source, tool, effect, learner, or credential authority.
"""

from __future__ import annotations

import re
from typing import Any, Mapping, Sequence

from .authoring import content_digest
from .autonomous_action_admission_persistence import (
    InMemoryAutonomousActionAdmissionLedger,
    validate_autonomous_action_admission_record,
)
from .autonomous_action_execution import AutonomousActionAdmission, admit_autonomous_action_plan
from .autonomous_action_plan import AutonomousActionPlan
from .domain_tools import AUTONOMOUS_DOMAIN_NAMES
from .errors import ArgumentError


AUTONOMOUS_ACTION_REVIEW_QUEUE_SCHEMA = "bioprism-python-autonomous-action-review-queue/0.1"
AUTONOMOUS_ACTION_REVIEW_ROW_SCHEMA = "bioprism-python-autonomous-action-review-row/0.1"
AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA = "bioprism-python-autonomous-action-dispatch-handoff/0.1"
AUTONOMOUS_ACTION_REVIEW_RETENTION = "metadata_only;operator_review_projection_and_digests;task_prompt_provider_connector_credential_and_effect_values_not_retained"
AUTONOMOUS_ACTION_REVIEW_AUTHORITY = "caller_operator_projection_only;authorization_is_external_and_not_verified_by_sdk"
AUTONOMOUS_ACTION_REVIEW_EXECUTION = "review_control_only;does_not_authorize_provider_source_tool_effect_or_credentials"
AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL = "never_returned"
AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES = (
    "credential_scope",
    "provider_or_source_approval",
    "tool_and_effect_authority",
    "evaluator_settlement",
)

_HANDOFF_KEYS = {
    "schema", "action_id", "record_digest", "plan_digest", "admission_digest", "plan", "admission",
    "selected_domains", "requested_domains", "cross_domain", "execution_path", "status", "downstream_gates",
    "authority", "retention", "execution", "secret_material", "handoff_digest",
}


def _fail(message: str) -> None:
    raise ArgumentError(f"autonomous action review controller {message}")


def _digest(name: str, value: Any) -> str:
    if not isinstance(value, str) or len(value) != 64 or any(character not in "0123456789abcdef" for character in value):
        _fail(f"{name} must be a lowercase SHA-256 digest")
    return value


def _domains(name: str, values: Sequence[str]) -> list[str]:
    if isinstance(values, (str, bytes, bytearray)) or not isinstance(values, Sequence) or not 1 <= len(values) <= len(AUTONOMOUS_DOMAIN_NAMES):
        _fail(f"{name} must contain one to twelve domains")
    normalized = [str(value) for value in values]
    if any(value not in AUTONOMOUS_DOMAIN_NAMES for value in normalized):
        _fail(f"{name} contains an unsupported domain")
    if len(set(normalized)) != len(normalized):
        _fail(f"{name} must contain unique domains")
    return normalized


def _action_id(value: Any) -> str:
    if not isinstance(value, str) or not 1 <= len(value) <= 256 or re.fullmatch(r"[A-Za-z0-9_.:+/-]+", value) is None:
        _fail("handoff action_id is invalid")
    return value


def _normalized(record: Mapping[str, Any]) -> tuple[dict[str, Any], AutonomousActionPlan, AutonomousActionAdmission]:
    normalized = validate_autonomous_action_admission_record(record)
    plan = AutonomousActionPlan.from_dict(normalized["plan"])
    admission = AutonomousActionAdmission.from_dict(normalized["admission"])
    return normalized, plan, admission


def _row(record: Mapping[str, Any]) -> dict[str, Any]:
    normalized, plan, admission = _normalized(record)
    return {
        "schema": AUTONOMOUS_ACTION_REVIEW_ROW_SCHEMA,
        "action_id": normalized["action_id"],
        "revision": normalized["revision"],
        "status": normalized["status"],
        "plan_digest": plan.plan_digest,
        "admission_digest": admission.admission_digest,
        "route_digest": plan.route_digest,
        "selected_domains": list(plan.selected_domains),
        "cross_domain": plan.cross_domain,
        "execution_path": admission.execution_path,
        "next_action": admission.next_action,
        "next_actions": list(admission.next_actions),
        "required_approvals": list(admission.required_approvals),
        "approved_approvals": list(admission.approved_approvals),
        "missing_approvals": list(admission.missing_approvals),
        "review_reasons": list(admission.review_reasons),
        "blocking_reasons": list(admission.blocking_reasons),
        "reviewer_digest": normalized["reviewer_digest"],
        "reason_digest": normalized["reason_digest"],
        "record_digest": normalized["record_digest"],
        "authority": AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
        "retention": AUTONOMOUS_ACTION_REVIEW_RETENTION,
        "execution": AUTONOMOUS_ACTION_REVIEW_EXECUTION,
        "secret_material": AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
    }


def _queue(rows: Sequence[Mapping[str, Any]]) -> dict[str, Any]:
    sorted_rows = sorted((dict(row) for row in rows), key=lambda row: row["action_id"])
    domain_counts = {domain: 0 for domain in AUTONOMOUS_DOMAIN_NAMES}
    for row in sorted_rows:
        for domain in row["selected_domains"]:
            domain_counts[domain] += 1
    body = {
        "schema": AUTONOMOUS_ACTION_REVIEW_QUEUE_SCHEMA,
        "rows": sorted_rows,
        "counts": {
            "total": len(sorted_rows),
            "pending_review": sum(row["status"] == "pending_review" for row in sorted_rows),
            "admitted": sum(row["status"] == "admitted" for row in sorted_rows),
            "blocked": sum(row["status"] == "blocked" for row in sorted_rows),
        },
        "domain_counts": domain_counts,
        "authority": AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
        "retention": AUTONOMOUS_ACTION_REVIEW_RETENTION,
        "execution": AUTONOMOUS_ACTION_REVIEW_EXECUTION,
        "secret_material": AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
    }
    return {**body, "queue_digest": content_digest(body)}


def validate_autonomous_action_dispatch_handoff(value: Mapping[str, Any]) -> dict[str, Any]:
    """Verify a handoff after it crosses persistence or process boundaries.

    This proves metadata continuity only. Reviewer authorization, credential readiness, provider
    availability, source truth, evaluator quality, and effect safety remain downstream gates.
    """

    if not isinstance(value, Mapping) or set(value) != _HANDOFF_KEYS:
        _fail("handoff contains unsupported or missing fields")
    if value["schema"] != AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA or value["authority"] != AUTONOMOUS_ACTION_REVIEW_AUTHORITY or value["retention"] != AUTONOMOUS_ACTION_REVIEW_RETENTION or value["execution"] != AUTONOMOUS_ACTION_REVIEW_EXECUTION or value["secret_material"] != AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL:
        _fail("handoff markers are invalid")
    try:
        plan = AutonomousActionPlan.from_dict(value["plan"])
        admission = AutonomousActionAdmission.from_dict(value["admission"])
    except Exception as error:
        raise ArgumentError("autonomous action review controller handoff plan or admission is invalid") from error
    record_digest = _digest("handoff record_digest", value["record_digest"])
    plan_digest = _digest("handoff plan_digest", value["plan_digest"])
    admission_digest = _digest("handoff admission_digest", value["admission_digest"])
    supplied_digest = _digest("handoff handoff_digest", value["handoff_digest"])
    if plan_digest != plan.plan_digest or admission_digest != admission.admission_digest or plan.plan_digest != admission.plan_digest:
        _fail("handoff plan or admission digest is inconsistent")
    if admission.status != "admitted":
        _fail("handoff admission is not admitted")
    selected = _domains("handoff selected_domains", value["selected_domains"])
    requested = _domains("handoff requested_domains", value["requested_domains"])
    if selected != list(plan.selected_domains) or selected != list(admission.selected_domains):
        _fail("handoff selected domains are inconsistent")
    if any(domain not in selected for domain in requested):
        _fail("handoff requested domains exceed the selected domains")
    if not isinstance(value["cross_domain"], bool) or value["cross_domain"] != plan.cross_domain:
        _fail("handoff cross-domain posture is inconsistent")
    if not isinstance(value["execution_path"], str) or value["execution_path"] != admission.execution_path:
        _fail("handoff execution path is inconsistent")
    if value["status"] != "ready_for_downstream_gates":
        _fail("handoff status is invalid")
    if value["downstream_gates"] != list(AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES):
        _fail("handoff downstream gates are invalid")
    body = {
        "schema": AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA,
        "action_id": _action_id(value["action_id"]),
        "record_digest": record_digest,
        "plan_digest": plan_digest,
        "admission_digest": admission_digest,
        "plan": plan.to_dict(),
        "admission": admission.to_dict(),
        "selected_domains": selected,
        "requested_domains": requested,
        "cross_domain": value["cross_domain"],
        "execution_path": value["execution_path"],
        "status": "ready_for_downstream_gates",
        "downstream_gates": list(AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES),
        "authority": AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
        "retention": AUTONOMOUS_ACTION_REVIEW_RETENTION,
        "execution": AUTONOMOUS_ACTION_REVIEW_EXECUTION,
        "secret_material": AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
    }
    expected = content_digest(body)
    if supplied_digest != expected:
        _fail("handoff digest does not match metadata")
    return {**body, "handoff_digest": expected}


class AutonomousActionAdmissionController:
    """Bounded operator projection over an action-admission ledger."""

    def __init__(self, ledger: InMemoryAutonomousActionAdmissionLedger) -> None:
        if not isinstance(ledger, InMemoryAutonomousActionAdmissionLedger):
            _fail("requires a typed action admission ledger")
        self.ledger = ledger

    def queue(self) -> dict[str, Any]:
        return _queue([_row(record) for record in self.ledger.list()])

    def get(self, action_id: str) -> dict[str, Any] | None:
        record = self.ledger.get(action_id)
        return None if record is None else _row(record)

    def submit(
        self,
        action_id: str,
        plan: AutonomousActionPlan | Mapping[str, Any],
        *,
        approvals: Mapping[str, bool] | None = None,
        reviewed: bool = False,
        authorization_digest: str | None = None,
        reason: str | None = None,
    ) -> dict[str, Any]:
        """Submit a provider-free plan to the durable review queue.

        Ordinary submissions stay pending until :meth:`review` is called. A caller may submit
        an already-admitted plan only when it also supplies the external authorization digest;
        that digest is recorded as review identity, never treated as a provider credential.
        """

        parsed = plan if isinstance(plan, AutonomousActionPlan) else AutonomousActionPlan.from_dict(plan)
        admission = admit_autonomous_action_plan(parsed, approvals=approvals, reviewed=reviewed)
        authorization = None if authorization_digest is None else _digest("authorization_digest", authorization_digest)
        if admission.status == "admitted" and authorization is None:
            _fail("an admitted submission requires an authorization_digest")
        record = self.ledger.submit(
            parsed,
            admission,
            action_id=action_id,
            reviewer_digest=authorization,
            reason=reason,
        )
        return _row(record)

    def review(
        self,
        action_id: str,
        *,
        authorization_digest: str,
        approvals: Mapping[str, bool] | None = None,
        reviewed: bool = False,
        reason: str | None = None,
        expected_record_digest: str | None = None,
    ) -> dict[str, Any]:
        authorization = _digest("authorization_digest", authorization_digest)
        record = self.ledger.review(
            action_id,
            approvals=approvals,
            reviewed=reviewed,
            reviewer_digest=authorization,
            reason=reason,
            expected_record_digest=expected_record_digest,
        )
        return _row(record)

    def dispatch_handoff(self, action_id: str, requested_domains: Sequence[str] | None = None) -> dict[str, Any]:
        record = self.ledger.get(action_id)
        if record is None:
            _fail("cannot create a handoff for an unknown action")
        normalized, plan, admission = _normalized(record)
        if normalized["status"] != "admitted" or admission.status != "admitted":
            _fail("action admission is not ready for downstream gates")
        selected = list(plan.selected_domains)
        requested = _domains("requested_domains", selected if requested_domains is None else requested_domains)
        missing = sorted(set(requested) - set(selected))
        if missing:
            _fail("requested domains are outside the admitted plan: " + ", ".join(missing))
        body = {
            "schema": AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA,
            "action_id": normalized["action_id"],
            "record_digest": normalized["record_digest"],
            "plan_digest": plan.plan_digest,
            "admission_digest": admission.admission_digest,
            "plan": plan.to_dict(),
            "admission": admission.to_dict(),
            "selected_domains": selected,
            "requested_domains": requested,
            "cross_domain": plan.cross_domain,
            "execution_path": admission.execution_path,
            "status": "ready_for_downstream_gates",
            "downstream_gates": list(AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES),
            "authority": AUTONOMOUS_ACTION_REVIEW_AUTHORITY,
            "retention": AUTONOMOUS_ACTION_REVIEW_RETENTION,
            "execution": AUTONOMOUS_ACTION_REVIEW_EXECUTION,
            "secret_material": AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL,
        }
        return validate_autonomous_action_dispatch_handoff({**body, "handoff_digest": content_digest(body)})


__all__ = [
    "AUTONOMOUS_ACTION_REVIEW_QUEUE_SCHEMA",
    "AUTONOMOUS_ACTION_REVIEW_ROW_SCHEMA",
    "AUTONOMOUS_ACTION_DISPATCH_HANDOFF_SCHEMA",
    "AUTONOMOUS_ACTION_REVIEW_RETENTION",
    "AUTONOMOUS_ACTION_REVIEW_AUTHORITY",
    "AUTONOMOUS_ACTION_REVIEW_EXECUTION",
    "AUTONOMOUS_ACTION_REVIEW_SECRET_MATERIAL",
    "AUTONOMOUS_ACTION_DISPATCH_DOWNSTREAM_GATES",
    "AutonomousActionAdmissionController",
    "validate_autonomous_action_dispatch_handoff",
]
